Internal/svTrainNN.ecl

```diff
@@ -0,0 +1,494 @@
+  * @param negsamp The number of negative samples to choose for each main word.
+  * @return weight updates as DATASET(SliceExt).  Note that these are additive 
+  *         changes to the weights, not the final weight values.
+  */
+EXPORT STREAMED DATASET(SliceExt) svTrainNN(STREAMED DATASET(SliceExt) wts,
+          STREAMED DATASET(trainingDat) train,
+          UNSIGNED4 slicesize,
+          UNSIGNED4 nWeights,
+          UNSIGNED4 numwords,
+          UNSIGNED4 dim,
+          UNSIGNED4 mbsize,
+          REAL lr,
+          UNSIGNED4 negsamp) := EMBED(C++ : activity)
+  #include <stdlib.h>
+  #include <stdio.h>
+  #include <stdint.h>
+  #include <string.h>
+  #include <math.h>
+  #include <assert.h>
+  /**
+  * Training Point Structure
+  */
+  struct tp
+  {
+    uint32_t main;
+    uint32_t contxCount;
+    uint32_t * contx;
+  };
+  /**
+    * Specialized Neural Network with 3 Layers.
+    * The input layer is the size of the vocabulary, the hidden layer is the
+    * size of the word vectors, and the output layer is the size of the vocabulary.
+    * Training data is presented as
+    * a tuple of a main word (id), and a list of context words (ids). Context
+    * is generally the words that were near or in the same sentence as the main
+    * word.
+    * The context words are simultaneously active at the input layer, and the
+    * main word is the training target at the output layer.
+    * The hidden layer weights represent the word vectors.  Since only
+    * The hidden layer is linear (no activation function)
+    * and the output layer uses a sigmoid activation function.
+    */
+  class NeurNet
+  {
+    double* w_; // Weights
+    double* wUpdates_; // Weight updates (output)
+    tp* t_; // Training Points
+    uint32_t tsize_; // Number of training recs
+    uint32_t wc_;  // Number of words in vocab
+    uint32_t dim_; // Length of word vector
+    double* hidden_;  // Forward pass results for hidden layer
+    uint32_t negSamp_; // Number of negative samples
+    double lr_; // Learning Rate
+    float loss_; // Accumulated loss
+    public:
+      NeurNet()
+      {
+      }
+      NeurNet(double * wts, double * wUpdates, uint32_t wc, uint32_t dim,
+              double lr, uint32_t negSamp)
+      {
+        w_ = wts;
+        wUpdates_ = wUpdates;
+        wc_ = wc;
+        dim_ = dim;
+        hidden_ = (double *)rtlMalloc(dim * sizeof(double));
+        lr_ = lr;
+        negSamp_ = negSamp;
+        loss_ = 0;
+      }
+      float getLoss()
+      {
+        return loss_;
+      }
+      // Convert triple index (l, j, i) to a flat index into weights
+      // l is the layer [0, 1]
+      // j,i is the weight from node j in layer l to node i in layer l+1
+      uint32_t toFlatIndx(uint16_t l, uint32_t j, uint32_t i)
+      {
+        uint32_t iSize, indx;
+        if (l == 0)
+          iSize = dim_;
+        else
+          iSize = wc_;
+        indx = l * dim_ * wc_;
```

Comment:
possibly better to initialise to 0 or dim_ * wc_ in the condition above - arguably clearer, and may make it more efficient when the function is inlined.  Also worth marking the function as const.  Some of the other functions could also be marked as const - e.g., getLoss().

Reply:
Could possibly have different functions for the different layers - since the layer is always a constant.

Reply:
deferred